metrics/perplexity/perplexity.py

```diff
@@ -0,0 +1,146 @@
+        646.74
+
+    Example 2:
+        >>> perplexity = datasets.load_metric("perplexity")
+        >>> input_texts = datasets.load_dataset("wikitext", "wikitext-2-raw-v1", "test") # doctest:+ELLIPSIS
+        [...]
+        >>> results = perplexity.compute(model_id='gpt2',
+        ...                              input_texts=input_texts,
+        ...                              stride=256)
+        >>> round(results["perplexity"], 2)
+        285.12
+
+"""
+
+
+@datasets.utils.file_utils.add_start_docstrings(_DESCRIPTION, _KWARGS_DESCRIPTION)
+class Perplexity(datasets.Metric):
+    def _info(self):
+        return datasets.MetricInfo(
+            description=_DESCRIPTION,
+            citation=_CITATION,
+            inputs_description=_KWARGS_DESCRIPTION,
+            features=datasets.Features(
+                {
+                    "input_texts": datasets.Value("string"),
+                }
+            ),
+            reference_urls=["https://huggingface.co/docs/transformers/perplexity"],
+        )
+
+    def _compute(self, input_texts, model_id, stride=512, device=None):
+
+        if device is not None:
+            assert device in ["gpu", "cpu"], "device should be either gpu or cpu."
+        else:
+            device = "gpu" if torch.cuda.is_available() else "cpu"
+
+        model = AutoModelForCausalLM.from_pretrained(model_id)
+        model = model.to(device)
+
+        tokenizer = AutoTokenizer.from_pretrained(model_id, pad_token="<PAD>")
+
+        encodings = tokenizer(input_texts, padding=True, return_tensors="pt", return_special_tokens_mask=True)
+
+        encoded_texts = encodings["input_ids"]
+        special_tokens_masks = encodings["special_tokens_mask"]
+
+        max_model_length = model.config.n_positions
+
+        ppls = []
+
+        for text_index in tqdm(range(0, len(encoded_texts))):
+            encoded_text = encoded_texts[text_index]
+            special_tokens_mask = special_tokens_masks[text_index]
+
+            encoded_text_length = len(encoded_text) - special_tokens_mask.sum()
+
+            nlls = []
+
+            target_index = 1
+            num_predictions = 0
+
+            while target_index < encoded_text_length:
+                start_index = max(0, target_index - (max_model_length - 1))
+
+                input_ids = encoded_text[start_index : target_index + 1]
+
+                target_ids = input_ids.clone()
+                target_ids[:-1] = -100
+
+                with torch.no_grad():
+                    outputs = model(input_ids, labels=target_ids)
+                    neg_log_likelihood = outputs[0]
+
+                nlls.append(neg_log_likelihood)
+
+                target_index += stride
+                num_predictions += 1
+
+            if len(nlls) > 0:
+                ppls.append(torch.exp(torch.stack(nlls).sum() / num_predictions))
+
+        ppl = torch.stack(ppls).sum() / len(encoded_texts)
+
+        return {
+            "perplexity": float(ppl),
+        }
```

Comment:
I'm wondering why returning a dict instead of a float. If finally a float is returned, docstring should be changed accordingly.

CC: @lhoestq

Reply:
For consistency with the other metrics. Certain metrics outputs several scores (e.g. rouge), so we output a dictionary. Even `accuracy` returns a dictionary.

Note that this can change if we build a new lib for metrics. Especially since Keras metrics expect one single number.
cc @lvwerra